src/model/level/internal.rs

```diff
@@ -0,0 +1,332 @@
+            "4" => LevelLength::ExtraLong,
+            int => LevelLength::Unknown(int.parse().map_err(D::Error::custom)?),
+        })
+    }
+}
+
+mod ten_bool {
+    use serde::{de::Error, Deserialize, Deserializer, Serializer};
+
+    pub fn serialize<S>(to_serialize: &bool, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match to_serialize {
+            true => serializer.serialize_str("10"),
+            false => serializer.serialize_str("0"),
+        }
+    }
+
+    pub fn deserialize<'de, D>(deserializer: D) -> Result<bool, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        match <&str>::deserialize(deserializer)? {
+            "10" => Ok(true),
+            "0" | "" => Ok(false),
+            _ => Err(D::Error::custom("expected '10', '0' or the empty string")),
+        }
+    }
+}
+
+impl LevelRating {
+    fn from_response_value(value: i32) -> LevelRating {
+        match value {
+            0 => LevelRating::NotAvailable,
+            10 => LevelRating::Easy,
+            20 => LevelRating::Normal,
+            30 => LevelRating::Hard,
+            40 => LevelRating::Harder,
+            50 => LevelRating::Insane,
+            _ => LevelRating::Unknown(value),
+        }
+    }
+
+    fn from_request_value(value: i32) -> LevelRating {
+        match value {
+            -3 => LevelRating::Auto,
+            -2 => LevelRating::Demon(DemonRating::Unknown(-1)), // The value doesn't matter, since setting the request field "rating" to
+            // -2 means "search for any demon, regardless of difficulty"
+            -1 => LevelRating::NotAvailable,
+            1 => LevelRating::Easy,
+            2 => LevelRating::Normal,
+            3 => LevelRating::Hard,
+            4 => LevelRating::Harder,
+            5 => LevelRating::Insane,
+            _ => LevelRating::Unknown(value),
+        }
+    }
+
+    fn into_response_value(self) -> i32 {
+        match self {
+            LevelRating::Unknown(value) => value,
+            LevelRating::NotAvailable => 0,
+            LevelRating::Easy => 20,
+            LevelRating::Normal => 30,
+            LevelRating::Hard => 40,
+            LevelRating::Harder => 50,
+            LevelRating::Insane => 60,
+            LevelRating::Demon(demon_rating) => demon_rating.into_response_value(),
+            _ => panic!("got {:?}, please handle before calling this function", self),
+        }
+    }
+
+    fn into_request_value(self) -> i32 {
+        match self {
+            LevelRating::Unknown(value) => value,
+            LevelRating::NotAvailable => -1,
+            LevelRating::Auto => -3,
+            LevelRating::Easy => 1,
+            LevelRating::Normal => 2,
+            LevelRating::Hard => 3,
+            LevelRating::Harder => 4,
+            LevelRating::Insane => 5,
+            LevelRating::Demon(_) => -2,
+        }
+    }
+}
+#[derive(Debug, Serialize, Deserialize)]
+pub struct InternalLevel<'a> {
```

Comment:
Maybe we could refer the values to the existing GD Docs? The website I found was http://docs.gdprogra.me/#/resources/server/level (Still lacks https), we could add this for now and change it later in case the docs project gets rewritten or rehosted elsewhere.

Reply:
Yee I looked that over to see if we could fill in some gaps in our documentation from there, but from what I've gathered our doc comments are already more accurate than what's given on there.

Reply:
Doc comments on the structs in the public API that is